src/wallet/reserves.rs

```diff
@@ -0,0 +1,561 @@
+    ChallengeInputMismatch,
+    /// Found an input other than the challenge which is not spendable. Holds the position of the input.
+    NonSpendableInput(usize),
+    /// Found an input that has no signature at position
+    NotSignedInput(usize),
+    /// Found an input with an unsupported SIGHASH type at position
+    UnsupportedSighashType(usize),
+    /// Found an input that is neither witness nor legacy at position
+    NeitherWitnessNorLegacy(usize),
+    /// Signature validation failed
+    SignatureValidation(usize, String),
+    /// The output is not valid
+    InvalidOutput,
+    /// Input and output values are not equal, implying a miner fee
+    InAndOutValueNotEqual,
+    /// No matching outpoing found
+    OutpointNotFound(usize),
+}
+
+impl<B, D> ProofOfReserves for Wallet<B, D>
+where
+    D: BatchDatabase,
+{
+    fn create_proof(&self, message: &str) -> Result<PSBT, Error> {
+        let challenge_txin = challenge_txin(message);
+        let challenge_psbt_inp = Input {
+            witness_utxo: Some(TxOut {
+                value: 0,
+                script_pubkey: Builder::new().push_opcode(opcodes::OP_TRUE).into_script(),
+            }),
+            final_script_sig: Some(Script::default()), // "finalize" the input with an empty scriptSig
+            ..Default::default()
+        };
+
+        let pkh = PubkeyHash::from_hash(hash160::Hash::hash(&[0]));
+        let out_script_unspendable = bitcoin::Address {
+            payload: Payload::PubkeyHash(pkh),
+            network: self.network,
+        }
+        .script_pubkey();
+
+        let mut builder = self.build_tx();
+        builder
+            .drain_wallet()
+            .add_foreign_utxo(challenge_txin.previous_output, challenge_psbt_inp, 42)?
+            .fee_absolute(0)
+            .only_witness_utxo()
+            .set_single_recipient(out_script_unspendable)
+            .ordering(TxOrdering::Untouched);
+        let (psbt, _details) = builder.finish().unwrap();
+
+        Ok(psbt)
+    }
+
+    fn verify_proof(&self, psbt: &PSBT, message: &str) -> Result<u64, Error> {
+        // verify the proof UTXOs are still spendable
+        let outpoints = self
+            .list_unspent()?
+            .iter()
+            .map(|utxo| (utxo.outpoint, utxo.txout.clone()))
+            .collect();
+
+        verify_proof(psbt, message, outpoints, self.network)
+    }
+}
+
+/// Make sure this is a proof, and not a spendable transaction.
+/// Make sure the proof is valid.
+/// Currently proofs can only be validated against the tip of the chain.
+/// If some of the UTXOs in the proof were spent in the meantime, the proof will fail.
+/// We can currently not validate whether it was valid at a certain block height.
+/// Returns the spendable amount of the proof.
+pub fn verify_proof(
+    psbt: &PSBT,
+    message: &str,
+    outpoints: Vec<(OutPoint, TxOut)>,
+    network: Network,
+) -> Result<u64, Error> {
+    let tx = psbt.clone().extract_tx();
+
```

Comment:
I'm not sure if it would make sense to copy the outpoints into a sorted container for faster lookups here.